presto-orc/src/main/java/com/facebook/presto/orc/writer/SliceDictionaryColumnWriter.java

```diff
@@ -0,0 +1,463 @@
+
+        if (directEncoded) {
+            return directColumnWriter.writeIndexStreams(outputStream, metadataWriter);
+        }
+
+        ImmutableList.Builder<RowGroupIndex> rowGroupIndexes = ImmutableList.builder();
+
+        List<LongStreamCheckpoint> dataCheckpoints = dataStream.getCheckpoints();
+        Optional<List<BooleanStreamCheckpoint>> presentCheckpoints = presentStream.getCheckpoints();
+        for (int i = 0; i < rowGroups.size(); i++) {
+            int groupId = i;
+            ColumnStatistics columnStatistics = rowGroups.get(groupId).getColumnStatistics();
+            LongStreamCheckpoint dataCheckpoint = dataCheckpoints.get(groupId);
+            Optional<BooleanStreamCheckpoint> presentCheckpoint = presentCheckpoints.map(checkpoints -> checkpoints.get(groupId));
+            List<Integer> positions = createSliceColumnPositionList(compression != NONE, dataCheckpoint, presentCheckpoint);
+            rowGroupIndexes.add(new RowGroupIndex(positions, columnStatistics));
+        }
+
+        int length = metadataWriter.writeRowIndexes(outputStream, rowGroupIndexes.build());
+        return ImmutableList.of(new Stream(column, StreamKind.ROW_INDEX, length, false));
+    }
+
+    private static List<Integer> createSliceColumnPositionList(
+            boolean compressed,
+            LongStreamCheckpoint dataCheckpoint,
+            Optional<BooleanStreamCheckpoint> presentCheckpoint)
+    {
+        ImmutableList.Builder<Integer> positionList = ImmutableList.builder();
+        presentCheckpoint.ifPresent(booleanStreamCheckpoint -> positionList.addAll(booleanStreamCheckpoint.toPositionList(compressed)));
+        positionList.addAll(dataCheckpoint.toPositionList(compressed));
+        return positionList.build();
+    }
+
+    @Override
+    public List<Stream> writeDataStreams(SliceOutput outputStream)
+            throws IOException
+    {
+        checkState(closed);
+
+        if (directEncoded) {
+            return directColumnWriter.writeDataStreams(outputStream);
+        }
+
+        // actually write data
+        ImmutableList.Builder<Stream> dataStreams = ImmutableList.builder();
+
+        presentStream.writeDataStreams(column, outputStream).ifPresent(dataStreams::add);
+        dataStream.writeDataStreams(column, outputStream).ifPresent(dataStreams::add);
+        dictionaryLengthStream.writeDataStreams(column, outputStream).ifPresent(dataStreams::add);
+        dictionaryDataStream.writeDataStreams(column, outputStream).ifPresent(dataStreams::add);
+        return dataStreams.build();
+    }
+
+    @Override
+    public long getBufferedBytes()
+    {
+        if (directEncoded) {
+            return directColumnWriter.getBufferedBytes();
+        }
+        // for dictionary columns we report the data we expect to write to the output stream
+        int indexBytes = estimateIndexBytesPerValue(dictionary.getEntryCount()) * getNonNullValueCount();
+        return indexBytes + getDictionaryBytes();
```

Comment:
Can this method be called after `close`? If not, add an assert. If so, there will be a problem because you have cleared 'dictionary'.